src/UnitCell.cpp

```diff
@@ -9,261 +9,245 @@
 }
 
-UnitCell::UnitCell(): UnitCell(0) {
-    shape_ = INFINITE;
+static bool is_roughly_zero(double value) {
+    // We think that 0.00001 is close enough to 0
+    return fabs(value) < 1e-5;
 }
 
-UnitCell::UnitCell(double a): UnitCell(a, a, a) {}
-
-UnitCell::UnitCell(double a, double b, double c): UnitCell(a, b, c, 90, 90, 90) {}
-
-UnitCell::UnitCell(double a, double b, double c, double alpha, double beta, double gamma):
-    h_(Matrix3D::unit()),
-    h_inv_(Matrix3D::unit()),
-    a_(a),
-    b_(b),
-    c_(c),
-    alpha_(alpha),
-    beta_(beta),
-    gamma_(gamma)
-{
-    auto is_zero = [](double length) {
-        // We think that 0.00001 is close enough to 0
-        return fabs(length) < 1e-5;
-    };
-    auto is_90 = [](double angle) {
-        // We think that 89.999° is close enough to 90°
-        return fabs(angle - 90.0) < 1e-3;
-    };
-    if (is_zero(a_) && is_zero(b_) && is_zero(c_)) {
-        shape_ = INFINITE;
-        a_ = b_ = c_ = 0;
-        alpha_ = beta_ = gamma_ = 90;
-    } else if (is_90(alpha_) && is_90(beta_) && is_90(gamma_)) {
-        shape_ = ORTHORHOMBIC;
-        // Make sure alpha/beta/gamma are actually 90°, so that the matrix
-        // update below does not create a non diagonal matrix.
-        alpha_ = 90;
-        beta_ = 90;
-        gamma_ = 90;
-    } else {
-        shape_ = TRICLINIC;
-    }
-    update_matrix();
+static bool is_roughly_90(double value) {
+    // We think that 89.999° is close enough to 90°
+    return fabs(value - 90.0) < 1e-3;
 }
 
-UnitCell::UnitCell(const Matrix3D& matrix): h_(Matrix3D::unit()), h_inv_(Matrix3D::unit()) {
-    if (matrix[1][0] != 0 || matrix[2][0] != 0 || matrix[2][1] != 0) {
-        throw error("the matrix supplied to UnitCell is not an upper triangular matrix");
-    }
+static bool is_diagonal(const Matrix3D& matrix) {
+    return is_roughly_zero(matrix[1][0]) && is_roughly_zero(matrix[2][0]) &&
+           is_roughly_zero(matrix[0][1]) && is_roughly_zero(matrix[2][1]) &&
+           is_roughly_zero(matrix[0][2]) && is_roughly_zero(matrix[1][2]);
+}
 
-    if (matrix[0][0] == 0 && matrix[1][1] == 0 && matrix[2][2] == 0 &&
-        matrix[0][1] == 0 && matrix[0][2] == 0 && matrix[1][2] == 0) {
+static bool is_upper_triangular(const Matrix3D& matrix) {
+    return is_roughly_zero(matrix[1][0]) && is_roughly_zero(matrix[2][0]) &&
+           is_roughly_zero(matrix[2][1]);
+}
 
-        shape_ = INFINITE;
-        a_ = b_ = c_ = 0;
-        alpha_ = beta_ = gamma_ = 90.0;
+static Matrix3D cell_matrix_from_lenths_angles(Vector3D lengths, Vector3D angles) {
+    if (lengths[0] < 0 || lengths[1] < 0 || lengths[2] < 0) {
+        throw error("can not create a unit cell with negative lengths");
+    }
 
-        update_matrix();
+    if (lengths != Vector3D(0, 0, 0) && (is_roughly_zero(lengths[0]) || is_roughly_zero(lengths[1]) || is_roughly_zero(lengths[2]))) {
+        warning("", "trying to create a unit cell with only one or two lengths set to zero");
```

Comment:
@ezavod: this is a warning instead of an error in part because of the `scaled_traj.nc` test file introduced in https://github.com/chemfiles/tests-data/commit/e1c6a19ad774a4fc7c3e55e59c30834bdcf7066e, which have cell lengths of `60.9682, 60.9682, 0`. Is this by design?

Reply:
I needed scaling for some very antiquated script a while ago. Seems to be a problem with that script that I didn't thought about at that time because I replaced the box dimensions for subsequent simulation.
Should I update the test files to some reasonable value so that one can have an error instead?

Reply:
I am not sure we want this to be an error: on one hand such cells do not make much sense, on the other hand they might be used to indicate partially periodic systems (for scaled_traj.nc, the system would only be periodic along x and y). Do you have a stronger opinion?

Reply:
Ok, let's keep it like this =)